### PR TITLE
feat: add support for installing instead of symlinking `file:` dependencies

### PR DESCRIPTION
**Pull request by nlf**
this adds an `installLinks` flag to arborist, which when set instructs arborist to pack and extract `file:` links rather than creating a symlink to their directory. this also causes us to reify the dependencies of the directory.

~a follow up pull request will add the config definition and documentation for the flag.~ i changed my mind about a separate pull request, i'll be adding some commits to this one instead.

for #2339
for npm/rfcs#150

### CODE REVIEW

File: workspaces/arborist/lib/arborist/build-ideal-tree.js [hunk 1]
```diff
@@ -1268,17 +1274,18 @@ This is a one-time fix-up, please be patient...
...
     // Don't bother to load the manifest for link deps, because the target
     // might be within another package that doesn't exist yet.
-    const { legacyPeerDeps } = this
+    const { installLinks, legacyPeerDeps } = this
+    const isWorkspace = this.idealTree.workspaces && this.idealTree.workspaces.has(spec.name)
 
-    // spec is a directory, link it
-    if (spec.type === 'directory') {
+    // spec is a directory, link it unless installLinks is set or it's a workspace
+    if (spec.type === 'directory' && (isWorkspace || !installLinks)) {
```
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/build-ideal-tree.js:1281]: So workspaces still explicitly don't get installed with this flag?  I *think* this is right but wanna be sure we have really thought about this.  What if we did install workspaces?  Would that make their install scripts run? Would that mean the workspaces themselves don't have `node_modules` folders?

Comment (nlf) [reply workspaces/arborist/lib/arborist/build-ideal-tree.js:1281]: correct, workspaces will always be symlinks with this change. it mostly just felt weird to install them. they live within the same directory tree as the root project, so we already reify their dependencies anyway. didn't feel like there was much to be gained by packing and installing them

**Review by wraithgar — APPROVED** [review]

**Review by ljharb — APPROVED** [review]
I'd really prefer if `link:` is for symlinks and `file:` is for copying files, like it used to be before npm 5 broke it, but another way to get file copying is still great

**Comment by nlf** [comment]
> I'd really prefer if `link:` is for symlinks and `file:` is for copying files, like it used to be before npm 5 broke it, but another way to get file copying is still great

that would be great, but also a breaking change. i'd personally really like to see npm@9 go back to pack+install for `file:` deps and require the use of `npm link` to create a link. this is an interim feature to unblock folks who depend on the old behavior without breaking folks who have already adjusted to the new.

**Comment by lukekarrys** [comment]
This looks great! 🎉 

@nlf Does it close either #2239 or https://github.com/npm/rfcs/pull/150?

File: tap-snapshots/test/lib/commands/ls.js.test.cjs [hunk 2]
```diff
@@ -204,21 +187,6 @@ exports[`test/lib/commands/ls.js TAP ls --parseable --long with extraneous deps
...
-`
-
-exports[`test/lib/commands/ls.js TAP ls --parseable --only=prod > should output tree containing only prod deps 1`] = `
-{CWD}/tap-testdir-ls-ls---parseable---only-prod
-{CWD}/tap-testdir-ls-ls---parseable---only-prod/node_modules/chai
-{CWD}/tap-testdir-ls-ls---parseable---only-prod/node_modules/optional-dep
-{CWD}/tap-testdir-ls-ls---parseable---only-prod/node_modules/prod-dep
-{CWD}/tap-testdir-ls-ls---parseable---only-prod/node_modules/prod-dep/node_modules/dog
-`
-
```
Comment (ruyadorno) [inline tap-snapshots/test/lib/commands/ls.js.test.cjs:221]: may land first here: https://github.com/npm/cli/pull/4776

**Comment by nlf** [comment]
> @nlf Does it close either #2239 or [npm/rfcs#150](https://github.com/npm/rfcs/pull/150)?

in theory, yes, but i didn't want it to automatically close the issues when it lands so i can follow up in both places

File: tap-snapshots/test/lib/commands/ls.js.test.cjs (hunk 2 above)
Comment (nlf) [reply tap-snapshots/test/lib/commands/ls.js.test.cjs:221]: good spot! i'll rebase this if #4776 lands first

**Comment by ThisIsMissEm** [comment]
Also perhaps related to @ljharb's comment: https://github.com/npm/feedback/discussions/667

**Pull request merged**